src/Control/FunFlow/Exec/Redis.hs

```diff
@@ -16,78 +17,225 @@ import Control.Monad.State.Strict
+  deriving (Generic, Show)
+
+data Job a = Job
+  { jobId :: JobId
+  , taskName :: T.Text
+  , jobStatus :: JobStatus
+  , jobError :: Maybe String
+  , argument :: a
+  } deriving Generic
+
+instance Store JobStatus
+instance Store a => Store (Job a)
+
+runRFlow :: R.Connection -> RFlowM a -> IO (Either String a)
+runRFlow conn mx = do
+  R.runRedis conn $ evalStateT (runExceptT mx) ("", conn)
+
+redis ::  R.Redis (Either R.Reply a) -> RFlowM a
+redis r = do ex <- lift $ lift r
+             case ex of
+               Left rply -> throwError $ "redis: "++show rply
+               Right x -> return x
 
 nameForKey :: T.Text -> RFlowM ByteString
 nameForKey k = do
-  (_,_,ns) <- get
+  (ns,_) <- get
   return $ ns <> "_" <> DTE.encodeUtf8 k
 
-rfresh :: RFlowM T.Text
-rfresh = do
-  (fs, ctx, ns) <- get
-  let (f,nfs) = popFreshers fs
-  put (nfs, ctx, ns)
-  return f
-
-fetchSym ::  Store a => T.Text -> RFlowM (Maybe a)
-fetchSym k = do
-  mv <- RFlowM . lift . R.get  =<< nameForKey k
-  let process (Left _) = return Nothing
-      process (Right x) = do
-        putLocal k x
-        return $ Just x
+fresh :: RFlowM T.Text
+fresh = do
+  n <- redis $ R.incr "fffresh"
+  return $ T.pack $ show (n::Integer)
+
+lookupSym ::  Store a => T.Text -> RFlowM (Maybe a)
+lookupSym k = do
+  mv <- redis . R.get  =<< nameForKey k
   case mv of
-    Right (Just v) -> process $ decode v
+    Just v -> return $ eitherToMaybe $ decode v
     _ -> return Nothing
 
 eitherToMaybe :: Either a b -> Maybe b
 eitherToMaybe (Left _) = Nothing
 eitherToMaybe (Right x) = Just x
 
-
-rlookupSym :: Store a => T.Text -> RFlowM (Maybe a)
-rlookupSym k = do
-  (_, ctx, _) <- get
-  case M.lookup k ctx of
-    Nothing -> fetchSym k
-    Just yv -> case decode yv of
-                Right y -> return $ Just y
-                Left _ -> fetchSym k
-
-rputSym :: Store a => T.Text -> a -> RFlowM ()
-rputSym k x = do
-  _ <- RFlowM . lift . (`R.set` (encode x)) =<< nameForKey k
-  putLocal k x
-
-putLocal :: Store a => T.Text -> a -> RFlowM ()
-putLocal k x = do
-  (fs, ctx, ns) <- get
-  put $ (fs, M.insert k (encode x) ctx, ns)
-
-{-
-runTillDone :: Flow a b -> a -> IO b
-runTillDone f x = go M.empty where
-  go st0 = do
-    ey <- resumeFlow f x st0
-    case ey of
-      Right y -> return y
-      Left (err, st1) -> do putStrLn $ "Flow failed with "++err
-                            go st1
-
-resumeFlow :: Flow a b -> a -> PureCtx -> IO (Either (String, PureCtx) b)
-resumeFlow f ini ctx = do
-  (ex, st) <- runStateT (proceedFlow f ini) (initFreshers, ctx)
-  case ex of
-    Left err -> return $ Left (err,snd st)
-    Right x ->  return $ Right x
--}
+fromRight :: Show a => Either a b -> b
+fromRight (Right x) = x
+fromRight (Left e) = error $ "fromRight: Left "++show e
+
+putSym :: Store a => T.Text -> a -> RFlowM ()
+putSym k x = do
+  _ <- redis .  (`R.set` (encode x)) =<< nameForKey k
+  return ()
+
+
+puttingSym :: Store a => T.Text -> a -> RFlowM a
```

Comment:
Convention is usually to have `foo` and `foo_` where one approach discards results - e.g. `forM`, `forM_` etc. Might be good to stick with that?